src/sentry/eventstore/models.py

```diff
@@ -157,45 +180,354 @@ def get_minimal_user(self):
+        information if available.  Grouping hashes will take into account
+        fingerprinting and checksums.
+        """
+        # If we have hashes stored in the data we use them, otherwise we
+        # fall back to generating new ones from the data.  We can only use
+        # this if we do not force a different config.
+        if force_config is None:
+            hashes = self.data.get("hashes")
+            if hashes is not None:
+                return hashes
+
+        return filter(
+            None, [x.get_hash() for x in self.get_grouping_variants(force_config).values()]
+        )
+
+    def get_grouping_variants(self, force_config=None, normalize_stacktraces=False):
+        """
+        This is similar to `get_hashes` but will instead return the
+        grouping components for each variant in a dictionary.
+
+        If `normalize_stacktraces` is set to `True` then the event data will be
+        modified for `in_app` in addition to event variants being created.  This
+        means that after calling that function the event data has been modified
+        in place.
+        """
+        from sentry.grouping.api import get_grouping_variants_for_event, load_grouping_config
+        from sentry.stacktraces.processing import normalize_stacktraces_for_grouping
+
+        # Forcing configs has two separate modes.  One is where just the
+        # config ID is given in which case it's merged with the stored or
+        # default config dictionary
+        if force_config is not None:
+            if isinstance(force_config, six.string_types):
+                stored_config = self.get_grouping_config()
+                config = dict(stored_config)
+                config["id"] = force_config
+            else:
+                config = force_config
+
+        # Otherwise we just use the same grouping config as stored.  if
+        # this is None the `get_grouping_variants_for_event` will fill in
+        # the default.
+        else:
+            config = self.data.get("grouping_config")
+
+        config = load_grouping_config(config)
+        if normalize_stacktraces:
+            normalize_stacktraces_for_grouping(self.data, config)
+
+        return get_grouping_variants_for_event(self, config)
+
+    def get_primary_hash(self):
+        # TODO: This *might* need to be protected from an IndexError?
+        return self.get_hashes()[0]
+
+    @property
+    def real_message(self):
+        # XXX(mitsuhiko): this is a transitional attribute that should be
+        # removed.  `message` will be renamed to `search_message` and this
+        # will become `message`.
+        return (
+            get_path(self.data, "logentry", "formatted")
+            or get_path(self.data, "logentry", "message")
+            or ""
+        )
+
+    @property
+    def organization(self):
+        return self.project.organization
+
+    @property
+    def version(self):
+        return self.data.get("version", "5")
+
+    def get_tag(self, key):
```

Comment:
Let's move this close to the tags property. This class has a lot of methods and properties, let's try at least to group them logically.